### PR TITLE
Add targeted endpoints

### PR DESCRIPTION
**Pull request by phlptp**
### Summary

If merged this pull request will addTargettedEndpoints and make some major changes to the messageAPI

### Proposed changes

- change to send, sendTo, sendAt, sendToAt for endpoints
- allow endpoints to have source and destination targets

### CODE REVIEW

**Comment by phlptp** [comment]
@nightlark  got everything but the travis builds from passing,  probably ready for you to review

File: src/helics/shared_api_library/MessageFederate.h [hunk 1]
```diff
@@ -65,6 +65,55 @@ HELICS_EXPORT helics_endpoint helicsFederateRegisterGlobalEndpoint(helics_federa
...
+ * @forcpponly
+ * @param[in,out] err A pointer to an error object for catching errors.
+ * @endforcpponly
+ *
+ * @return An object containing the endpoint.
+ * @forcpponly
+ *         nullptr on failure.
+ * @endforcpponly
+ */
+HELICS_EXPORT helics_endpoint helicsFederateRegisterTargettedEndpoint(helics_federate fed,
```
Comment (nightlark) [inline src/helics/shared_api_library/MessageFederate.h:88]:
```suggestion
HELICS_EXPORT helics_endpoint helicsFederateRegisterTargetedEndpoint(helics_federate fed,
```

File: src/helics/shared_api_library/MessageFederate.h [hunk 2]
```diff
@@ -65,6 +65,55 @@ HELICS_EXPORT helics_endpoint helicsFederateRegisterGlobalEndpoint(helics_federa
...
+ * @param type A string describing the expected type of the publication (may be NULL).
+ * @forcpponly
+ * @param[in,out] err A pointer to an error object for catching errors.
+ * @endforcpponly
+ * @return An object containing the endpoint.
+ * @forcpponly
+ *         nullptr on failure.
+ * @endforcpponly
+ */
+HELICS_EXPORT helics_endpoint helicsFederateRegisterGlobalTargettedEndpoint(helics_federate fed,
```
Comment (nightlark) [inline src/helics/shared_api_library/MessageFederate.h:112]:
```suggestion
HELICS_EXPORT helics_endpoint helicsFederateRegisterGlobalTargetedEndpoint(helics_federate fed,
```

File: src/helics/shared_api_library/MessageFederateExport.cpp [hunk 3]
```diff
@@ -66,6 +66,27 @@ helics_endpoint helicsFederateRegisterEndpoint(helics_federate fed, const char*
     return nullptr;
 }
 
+helics_endpoint helicsFederateRegisterTargettedEndpoint(helics_federate fed, const char* name, const char* type, helics_error* err)
```
Comment (nightlark) [inline src/helics/shared_api_library/MessageFederateExport.cpp:69]:
```suggestion
helics_endpoint helicsFederateRegisterTargetedEndpoint(helics_federate fed, const char* name, const char* type, helics_error* err)
```

File: src/helics/shared_api_library/MessageFederateExport.cpp [hunk 4]
```diff
@@ -66,6 +66,27 @@ helics_endpoint helicsFederateRegisterEndpoint(helics_federate fed, const char*
...
+{
+    // now generate a generic endpoint
+    auto fedObj = getMessageFedSharedPtr(fed, err);
+    if (!fedObj) {
+        return nullptr;
+    }
+    try {
+        auto end = std::make_unique<helics::EndpointObject>();
+        end->endPtr = &fedObj->registerTargettedEndpoint(AS_STRING(name), AS_STRING(type));
```
Comment (nightlark) [inline src/helics/shared_api_library/MessageFederateExport.cpp:78]:
```suggestion
        end->endPtr = &fedObj->registerTargetedEndpoint(AS_STRING(name), AS_STRING(type));
```

File: src/helics/shared_api_library/MessageFederateExport.cpp [hunk 5]
```diff
@@ -86,6 +107,26 @@ helics_endpoint helicsFederateRegisterGlobalEndpoint(helics_federate fed, const
     return nullptr;
 }
 
+helics_endpoint helicsFederateRegisterGlobalTargettedEndpoint(helics_federate fed, const char* name, const char* type, helics_error* err)
```
Comment (nightlark) [inline src/helics/shared_api_library/MessageFederateExport.cpp:110]:
```suggestion
helics_endpoint helicsFederateRegisterGlobalTargetedEndpoint(helics_federate fed, const char* name, const char* type, helics_error* err)
```

File: src/helics/shared_api_library/MessageFederateExport.cpp [hunk 6]
```diff
@@ -86,6 +107,26 @@ helics_endpoint helicsFederateRegisterGlobalEndpoint(helics_federate fed, const
...
+{
+    // now generate a generic subscription
+    auto fedObj = getMessageFedSharedPtr(fed, err);
+    if (!fedObj) {
+        return nullptr;
+    }
+    try {
+        auto end = std::make_unique<helics::EndpointObject>();
+        end->endPtr = &fedObj->registerGlobalTargettedEndpoint(AS_STRING(name), AS_STRING(type));
```
Comment (nightlark) [inline src/helics/shared_api_library/MessageFederateExport.cpp:119]:
```suggestion
        end->endPtr = &fedObj->registerGlobalTargetedEndpoint(AS_STRING(name), AS_STRING(type));
```

**Review by nightlark — COMMENTED** [review]
Mostly docstrings. Main issue is spelling targeted/targetted, which I think comes down to american/british spelling.

**Review by nightlark — APPROVED** [review]

**Pull request merged**